expr: give a name to a tuple of columns

avikivity opened the pull request:
Right now, binary_operator::lhs is a variant<column_value,
std::vector<column_value>, token>. The role of the second branch
(a vector of column values) is to represent a tuple of columns
e.g. "WHERE (a, b, c) = ?"), but this is not clear from the type
name.

Inroduce a wrapper type around the vector, column_value_tuple, to
make it clear we're dealing with tuples of CQL references (a
column_value is really a column_ref, since it doesn't actually
contain any value).

nyh commented on cql3/expr/expression.hh:
```diff
@@ -67,6 +67,11 @@ struct column_value {
+/// A tuple of column selectors. Usually used for range constraints on clustering keys.
+struct column_value_tuple {
+    std::vector<column_value> elements;
+};
```
Do you forsee this type ever needing to be contain anything except these "elements"?
If not, maybe it should be just an alias (using column_value_tuple = std::vector<column_value>)?
Or inheritence: struct column_value_tuple : public std::vector<column_value> ?
It could save you a lot of places typing the word "elements".

avikivity replied:
I don't mind a little extra typing if it brings some clarity. And I say that as a slow typist.

avikivity replied:
Or rather, I type that as a slow typist.

A slow typist who likes strong types!

dekimir replied:
I think it's fine, just like `conjunction` has a single member `children`.

nyh commented:
> Inroduce a wrapper type around the vector, column_value_tuple, to
> make it clear we're dealing with tuples of CQL references (a
> column_value is really a column_ref, since it doesn't actually
> contain any value).

Hmm, the name "column_value" does seem to suggest it holds a value - not a name of a column... Maybe the name should be rethought? @dekimir ?

avikivity commented:
I'm happy to follow up with a rename to `column_ref` later.

dekimir commented on cql3/restrictions/multi_column_restriction.hh:
```diff
@@ -190,7 +190,7 @@ public:
         using namespace expr;
         expression = binary_operator{
-            std::vector<column_value>(_column_defs.cbegin(), _column_defs.cend()), oper_t::EQ, _value};
+            expr::column_value_tuple{std::vector<column_value>(_column_defs.cbegin(), _column_defs.cend())}, oper_t::EQ, _value};
```
Yuck!  There has to be a shorter way to express the simple idea of "tuple from these defs".  How about adding a constructor?

avikivity replied:
Added a constructor.

dekimir commented:
> > Inroduce a wrapper type around the vector, column_value_tuple, to
> > make it clear we're dealing with tuples of CQL references (a
> > column_value is really a column_ref, since it doesn't actually
> > contain any value).
> 
> Hmm, the name "column_value" does seem to suggest it holds a value - not a name of a column... Maybe the name should be rethought? @dekimir ?

It should definitely be rethought but as a part of a larger refactoring.  Note that a column_value may contain a subscript, except when it's in a tuple.

dekimir commented:
In fact, maybe the tuple variant of `expression` should be a vector of column_definitions?  Or even just a positive number, since it must always be a clustering-key prefix?

avikivity commented:
Update: added a column_value_tuple constructor to make creation of tuples cleaner.

dekimir commented on cql3/restrictions/multi_column_restriction.hh:
```diff
@@ -190,7 +190,7 @@ public:
         using namespace expr;
         expression = binary_operator{
-            std::vector<column_value>(_column_defs.cbegin(), _column_defs.cend()), oper_t::EQ, _value};
+            expr::column_value_tuple(_column_defs), oper_t::EQ, _value};
```
`expr::` is redundant.

avikivity replied:
removed

dekimir commented on cql3/restrictions/multi_column_restriction.hh:
```diff
@@ -332,7 +332,7 @@ public:
         using namespace expr;
         expression = binary_operator{
-            std::vector<column_value>(_column_defs.cbegin(), _column_defs.cend()),
+            expr::column_value_tuple(_column_defs),
```
`expr::` is redundant.

avikivity replied:
removed

dekimir commented on cql3/restrictions/multi_column_restriction.hh:
```diff
@@ -361,7 +361,7 @@ public:
         using namespace expr;
         expression = binary_operator{
-            std::vector<column_value>(_column_defs.cbegin(), _column_defs.cend()),
+            expr::column_value_tuple(_column_defs),
```
`expr::` is redundant.

avikivity commented:
> In fact, maybe the tuple variant of `expression` should be a vector of column_definitions? Or even just a positive number, since it must always be a clustering-key prefix?

It should be a tuple of expressions. And column_value should be replaced by a column_reference (a wrapper around column_def*) and a struct subscription_expression { expression container; expression subscript; }.

dekimir reviewed (approved):
LGTM, with a couple of nitpicks.

dekimir commented:
> > In fact, maybe the tuple variant of `expression` should be a vector of column_definitions? Or even just a positive number, since it must always be a clustering-key prefix?
> 
> It should be a tuple of expressions.

Does that mean all code should be modified to handle something like `(a[3], 4, b, true) <= ('x', 4, 3, false)`, on the off chance that CQL will one day allow you to express that?

avikivity commented:
> > > In fact, maybe the tuple variant of `expression` should be a vector of column_definitions? Or even just a positive number, since it must always be a clustering-key prefix?
> > 
> > 
> > It should be a tuple of expressions.
> 
> Does that mean all code should be modified to handle something like `(a[3], 4, b, true) <= ('x', 4, 3, false)`, on the off chance that CQL will one day allow you to express that?

No, and this adds the additional difficulty that the constraints of the compiler/evaluator must match the constraints of the grammar. So there is value in having the constraints in the type system, but I think it's too constraining.

dekimir commented:
> there is value in having the constraints in the type system, but I think it's too constraining.

Let's see if you still think that after you see what it takes to extract the clustering-slice bounds even with the present constraints.

Hint:
https://github.com/dekimir/scylla/blob/cbcc094abe10647e14bf22e729361fe510e619cd/cql3/restrictions/statement_restrictions.cc#L144-L157

and:
https://github.com/dekimir/scylla/blob/cbcc094abe10647e14bf22e729361fe510e619cd/cql3/restrictions/statement_restrictions.cc#L901-L965

avikivity commented:
Updated:
 - rebased (with new instances of std::vector<column_value> converted
 - removed redundant expr:: prefixes (though it could be argued they help cross-module understanding)

dekimir reviewed (approved):
LGTM

avikivity commented:
PiNg

avikivity commented:
@scylladb/scylla-maint pInG

psarna commented:
There's a clash with changes introduced in 0bb959e890 - please help.

avikivity commented:
Rebased, retested, repushed.

psarna commented:
Reread, reviewed, remerged (without collisions this time)
